src/algebra/lie_algebra.lean

```diff
@@ -0,0 +1,213 @@
+of endomorphisms of a module as well as of the algebra of square matrices over a commutative ring.
+
+## Notations
+
+We introduce the notation ⁅x, y⁆ for the Lie bracket. Note that these are the Unicode "square with
+quill" brackets rather than the usual square brackets.
+
+## Implementation notes
+
+Lie algebras are defined as modules with a compatible Lie ring structure, and thus are partially
+unbundled. Since they extend Lie rings, these are also partially unbundled.
+
+## References
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 1--3*][bourbaki1975]
+
+## Tags
+
+lie bracket, ring commutator, jacobi identity, lie ring, lie algebra
+-/
+
+universes u v
+
+/--
+A product operation, intended use in Lie algebras and similar structures.
+-/
+class has_bracket (L : Type v) := (bracket : L → L → L)
+
+notation `⁅`x`,` y`⁆` := has_bracket.bracket x y
+
+namespace ring_commutator
+
+variables {A : Type v} [ring A]
+
+/--
+The ring commutator captures the extent to which a ring is commutative. It is identically zero
+exactly when the ring is commutative.
+-/
+def commutator (x y : A) := x*y - y*x
+
+local notation `⁅`x`,` y`⁆` := commutator x y
+
+@[simp] lemma add_left (x y z : A) :
+  ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ :=
+by simp [commutator, right_distrib, left_distrib]
+
+@[simp] lemma add_right (x y z : A) :
+  ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆ :=
+by simp [commutator, right_distrib, left_distrib]
+
+@[simp] lemma alternate (x : A) :
+  ⁅x, x⁆ = 0 :=
+by simp [commutator]
+
+lemma jacobi (x y z : A) :
+  ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0 :=
+begin
+  unfold commutator,
+  repeat { rw mul_sub_left_distrib },
+  repeat { rw mul_sub_right_distrib },
+  repeat { rw add_sub },
+  repeat { rw ←sub_add },
+  repeat { rw ←mul_assoc },
+  have h : ∀ (x y z : A), x - y + z + y = x+z := by simp,
+  repeat { rw h },
+  simp,
+end
+
+end ring_commutator
+
+/--
+A Lie ring is an additive group with compatible product, known as the bracket, satisfying the
+Jacobi identity. The bracket is not associative unless it is identically zero.
+-/
+class lie_ring (L : Type v) [add_comm_group L] extends has_bracket L :=
+(add_lie : ∀ (x y z : L), ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆)
+(lie_add : ∀ (x y z : L), ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆)
+(lie_self : ∀ (x : L), ⁅x, x⁆ = 0)
+(jacobi : ∀ (x y z : L), ⁅x, ⁅y, z⁆⁆ + ⁅y, ⁅z, x⁆⁆ + ⁅z, ⁅x, y⁆⁆ = 0)
+
+section lie_ring
+
+variables {L : Type v} [add_comm_group L] [lie_ring L]
+
+@[simp] lemma add_lie (x y z : L) : ⁅x + y, z⁆ = ⁅x, z⁆ + ⁅y, z⁆ := lie_ring.add_lie x y z
+@[simp] lemma lie_add (x y z : L) : ⁅z, x + y⁆ = ⁅z, x⁆ + ⁅z, y⁆ := lie_ring.lie_add x y z
+@[simp] lemma lie_self (x : L) : ⁅x, x⁆ = 0 := lie_ring.lie_self x
+
+@[simp] lemma lie_skew (x y : L) :
+  -⁅y, x⁆ = ⁅x, y⁆ :=
+begin
+  symmetry,
+  rw [←sub_eq_zero_iff_eq, sub_neg_eq_add],
+  have H : ⁅x + y, x + y⁆ = 0, from lie_self _,
+  rw [add_lie, lie_add, lie_add, lie_self, lie_self] at H,
```

Comment:
As far as I can see, you don't need this line. The `simpa` on the next line should apply all those simp-lemmas at `H`...

Reply:
For some reason I can get away with replacing the line with just:
```
rw add_lie at H,
```
but this is necessary. Perhaps it first tries to simplify using `lie_self` and then has `0` so it's a dead end?

Reply:
Aha, that certainly makes sense.